File: src/Djio2.java
import java.util.LinkedList;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Scanner;

public class Djio2 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] A = new int[n];
        for (int i = 0; i < n; i++) {
            A[i] = sc.nextInt();
        }
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b - a);
        for (int i : A) {
            pq.add(i);
        }
        int time = 0;
        int lastEaten = -1;
        Queue<Integer> rest = new LinkedList<>();
        while (!pq.isEmpty() || !rest.isEmpty()) {
            if (!pq.isEmpty()) {
                if (pq.peek() != lastEaten) {
                    lastEaten = pq.peek();
                    pq.poll();
                    if (lastEaten > 1) {
                        rest.add(lastEaten - 1);
                    }
                } else {
                    rest.add(pq.peek());
                    pq.poll();
                }
            }
            if (pq.isEmpty()) {
                while (!rest.isEmpty()) {
                    pq.add(rest.poll());
                }
                lastEaten = -1;
            }
            time++;
        }
        System.out.println(time);
        sc.close();
    }
}

File: bin/Djio2.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int main() {
    int N;
    cin >> N;
    vector<int> A(N);
    priority_queue<pair<int, int>> pq;
    queue<pair<int, int>> rest;
    int time = 0, lastEaten = 0;

    for (int i = 0; i < N; i++) {
        cin >> A[i];
        if (A[i] > 0) {
            pq.push({A[i], i});
        }
    }

    while (!pq.empty() || !rest.empty()) {
        if (!pq.empty()) {
            if (pq.top().second != lastEaten) {
                lastEaten = pq.top().second;
                int remaining = pq.top().first - 1;
                pq.pop();
                if (remaining > 0) {
                    rest.push({remaining, lastEaten});
                }
            } else {
                rest.push(pq.top());
                pq.pop();
            }
        }
        if (pq.empty() && !rest.empty()) {
            while (!rest.empty()) {
                pq.push(rest.front());
                rest.pop();
            }
            lastEaten = -1;
        }
        time++;
    }

    cout << time << endl;

    return 0;
}
